src/thor/expansion_action.cc

```diff
@@ -78,43 +118,56 @@ std::string thor_worker_t::expansion(Api& request) {
   // a lambda that the path algorithm can call to add stuff to the dom
   // route and isochrone produce different GeoJSON properties
   std::string algo = "";
-  auto track_expansion = [&expansion, &opp_edges, &gen_factor, &skip_opps, &exp_props,
-                          &algo](baldr::GraphReader& reader, baldr::GraphId edgeid,
-                                 baldr::GraphId prev_edgeid, const char* algorithm = nullptr,
-                                 const Expansion_EdgeStatus status = Expansion_EdgeStatus_reached,
-                                 const float duration = 0.f, const uint32_t distance = 0,
-                                 const float cost = 0.f) {
-    algo = algorithm;
-
-    auto tile = reader.GetGraphTile(edgeid);
-    if (tile == nullptr) {
-      LOG_ERROR("thor_worker_t::expansion error, tile no longer available" +
-                std::to_string(edgeid.Tile_Base()));
-      return;
-    }
-    const auto* edge = tile->directededge(edgeid);
-    // unfortunately we have to call this before checking if we can skip
-    // else the tile could change underneath us when we get the opposing
-    auto shape = tile->edgeinfo(edge).shape();
-    auto names = tile->edgeinfo(edge).GetNames();
-    auto is_forward = edge->forward();
-
-    // if requested, skip this edge in case its opposite edge has been added
-    // before (i.e. lower cost) else add this edge's id to the lookup container
-    if (skip_opps) {
-      auto opp_edgeid = reader.GetOpposingEdgeId(edgeid, tile);
-      if (opp_edgeid && opp_edges.count(opp_edgeid))
-        return;
-      opp_edges.insert(edgeid);
-    }
-
-    if (!edge->forward())
-      std::reverse(shape.begin(), shape.end());
-    Polyline2<PointLL>::Generalize(shape, gen_factor, {}, false);
-
-    writeExpansionProgress(expansion, edgeid, prev_edgeid, shape, exp_props, status, duration,
-                           distance, cost);
-  };
+  auto track_expansion =
+      [&expansion, &opp_edges, &gen_factor, &skip_opps, &exp_props, &algo, &edge_state,
+       &dedupe](baldr::GraphReader& reader, baldr::GraphId edgeid, baldr::GraphId prev_edgeid,
+                const char* algorithm = nullptr,
+                const Expansion_EdgeStatus status = Expansion_EdgeStatus_reached,
+                const float duration = 0.f, const uint32_t distance = 0, const float cost = 0.f) {
+        algo = algorithm;
+
+        auto tile = reader.GetGraphTile(edgeid);
+        if (tile == nullptr) {
+          LOG_ERROR("thor_worker_t::expansion error, tile no longer available" +
+                    std::to_string(edgeid.Tile_Base()));
+          return;
+        }
+        const auto* edge = tile->directededge(edgeid);
+        // unfortunately we have to call this before checking if we can skip
+        // else the tile could change underneath us when we get the opposing
+        auto shape = tile->edgeinfo(edge).shape();
```

Comment:
you're saying that because of the call to `GetOpposingEdgeId` below? if so the simple solution would be to, inside of the `skip_opps` `if` block to add `auto opp_tile = tile;` and use `opp_tile` in the call to `GetOpposingEdgeId`. if i misunderstood this comment let me know!

Reply:
This is original comment from @nilsnolde actually, I haven't touch it, it just got marked by diff as I've clean up some unused code around those lines

Reply:
i mean would be good to move the grabbing of the shape below the `if` though in case that method bails before using it

Reply:
Good point, I've moved it and removed the original comment